File: 0006.zigzag-conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if ("".equals(s) || numRows <= 1) {
            return s;
        }
        char ca[] = s.toCharArray();
        char result[] = new char[ca.length];
        int resIndex = 0;
         // 每行下标间隔
        int inc = 2 * (numRows -1);
        int index ;
        int n = numRows -1;
        for (int i=0;i<numRows && i < ca.length;i++) {
            index = i;
            // 首行和最后一行两列之间没有中间元素
            if(i % n == 0) {
                while (index < ca.length) {
                    result[resIndex++] = ca[index];
                    index +=inc;
                }
            } else {
                // 直接输入每行第一个元素
                result[resIndex++] = ca[index];
                while (index < ca.length) {
                    // 获取右列下标索引
                    int right = index + inc;
                    // 两列中间元素下标为右边列 - (行号*2)
                    int left = right - 2 * i;
                    if (left < ca.length) {
                        result[resIndex++] = ca[left];
                    }
                    if (right < ca.length) {
                        result[resIndex++] = ca[right];
                    }
                    index = right;
                }
            }
        }
        return String.valueOf(result);
    }
}

File: 0006.zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        n = numRows-2
        m = numRows + n
        l, r = "", {}
        
        if n >= 0:
            pass
        else:
            return s
        
        for j in range(numRows):
            r[j] = ""
        
        for i in range(len(s)//m+1):
            si = s[m*i:m*(i+1)]
            if si:
                si1, si2 = si[:numRows], si[numRows:]
                for j in range(numRows):
                    if j in [0, numRows-1]:
                        r[j] = r[j] + (si1[j] if j < len(si1) else "")
                    else:
                        r[j] = r[j] + (si1[j] if j < len(si1) else "") + (si2[n-j] if n-j < len(si2) else "")
        for i in r.values():
            l += i
        return l
